<?php

declare(strict_types=1);

/*
 * #logic 做事不讲究逻辑，再努力也只是重复犯错
 * ## 何为相思：不删不聊不打扰，可否具体点：曾爱过。何为遗憾：你来我往皆过客，可否具体点：再无你。
 * ## 只要思想不滑稽，方法总比苦难多！
 * @version 1.0.0
 * @author @小小只^v^ <littlezov@qq.com>  littlezov@qq.com
 * @contact  littlezov@qq.com
 * @link     https://github.com/littlezo
 * @document https://github.com/littlezo/wiki
 * @license  https://github.com/littlezo/MozillaPublicLicense/blob/main/LICENSE
 *
 */

namespace littler;

use littler\exceptions\ApiException;
use littler\exceptions\FailedException;
use think\db\exception\DataNotFoundException;
use think\db\exception\ModelNotFoundException;
use think\exception\Handle;
use think\exception\HttpException;
use think\exception\HttpResponseException;
use think\exception\ValidateException;
use think\Response;
use Throwable;
use xiaodi\JWTAuth\Exception\JWTException;

class ExceptionHandle extends Handle
{
	/**
	 * 不需要记录信息（日志）的异常类列表.
	 * @var array
	 */
	protected $ignoreReport = [
		HttpException::class,
		HttpResponseException::class,
		ModelNotFoundException::class,
		DataNotFoundException::class,
		ValidateException::class,
	];

	/**
	 * 记录异常信息（包括日志或者其它方式记录）.
	 */
	public function report(Throwable $exception): void
	{
		// 使用内置的方式记录异常日志
		parent::report($exception);
	}

	/**
	 * Render an exception into an HTTP response.
	 *
	 * @param \think\Request $request
	 * @throws \Exception
	 */
	public function render($request, Throwable $e): Response
	{
		// 其他错误交给系统处理
		if ($e instanceof \Exception && ! $e instanceof ApiException && ! $e instanceof JWTException) {
			$e = new FailedException($e->getMessage(), 10005, $e);
		}
		return parent::render($request, $e);
	}
}
